src/Data/Unfoldable.purs

```diff
@@ -38,3 +39,58 @@ instance unfoldableArray :: Unfoldable Array where
           writeSTRef seed b2
           return false
     return arr))
+
+-- | Replicate a value some natural number of times.
+-- | For example:
+-- |
+-- | ~~~ purescript
+-- | replicate 2 "foo" == ["foo", "foo"] :: Array String
+-- | ~~~
+replicate :: forall f a. (Unfoldable f) => Int -> a -> f a
+replicate n v = unfoldr step n
+  where
+    step :: Int -> Maybe (Tuple a Int)
+    step i =
+      if i <= 0 then Nothing
+      else Just (Tuple v (i - 1))
+
+-- | Perform an Applicative action `n` times, and accumulate all the results.
+replicateA :: forall m f a. (Applicative m, Unfoldable f, Traversable f) =>
+              Int -> m a -> m (f a)
+replicateA n m = sequence (replicate n m)
+
+-- | The empty container - unfolded with zero iterations.
+-- | For example:
+-- |
+-- | ~~~ purescript
+-- | empty == [] :: forall a. Array a
+-- | ~~~
+empty :: forall f a. (Unfoldable f) => f a
+empty = unfoldr (const Nothing) unit
+
+-- | Contain a single value.
+-- | For example:
+-- |
+-- | ~~~ purescript
+-- | singleton "foo" == ["foo"] :: Array String
+-- | ~~~
+singleton :: forall f a. (Unfoldable f) => a -> f a
+singleton = replicate 1
+
+-- | Perform an Applicative action and contain the single result.
+singletonA :: forall m f a. (Applicative m, Unfoldable f, Traversable f) =>
+              m a -> m (f a)
+singletonA = replicateA 1
```

Comment:
Feels like this should be `map singleton` although we probably lack the laws to prove it. If it is, maybe it's simple enough to not warrant its own function here, especially given the relatively complex instance context.